File: reference.vb.net/Db4objects.Db4odoc/Metainfo/MetaInfoExample.vb
' Copyright (C) 2004 - 2007 db4objects Inc. http://www.db4o.com 
Imports System
Imports System.IO
Imports Db4objects.Db4o
Imports Db4objects.Db4o.Ext

Namespace Db4objects.Db4odoc.MetaInfo

    Public Class MetaInfoExample
        Private Const Db4oFileName As String = "reference.db4o"

        Public Shared Sub Main(ByVal args() As String)
            SetObjects()
            GetMetaObjects()
            GetMetaObjectsInfo()
        End Sub
        ' end Main

        Private Shared Sub SetObjects()
            File.Delete(Db4oFileName)
            Dim container As IObjectContainer = Db4oFactory.OpenFile(Db4oFileName)
            Try
                Dim car As Car = New Car("BMW", New Pilot("Rubens Barrichello"))
                container.Set(car)
                car = New Car("Ferrari", New Pilot("Michael Schumacher"))
                container.Set(car)
            Finally
                container.Close()
            End Try
        End Sub
        ' end SetObjects

        Private Shared Sub GetMetaObjects()
            Dim container As IObjectContainer = Db4oFactory.OpenFile(Db4oFileName)
            Try
                System.Console.WriteLine("Retrieve meta information for class: ")
                Dim sc As IStoredClass = container.Ext().StoredClass(GetType(Car))
                System.Console.WriteLine("Stored class:  " + sc.GetName())

                System.Console.WriteLine("Retrieve meta information for all classes in database: ")
                Dim sclasses() As IStoredClass = container.Ext().StoredClasses()
                Dim i As Integer
                For i = 0 To sclasses.Length - 1 Step i + 1
                    System.Console.WriteLine(sclasses(i).GetName())
                Next
            Finally
                container.Close()
            End Try
        End Sub
        ' end GetMetaObjects

        Private Shared Sub GetMetaObjectsInfo()
            Dim container As IObjectContainer = Db4oFactory.OpenFile(Db4oFileName)
            Try
                System.Console.WriteLine("Retrieve meta information for field: ")
                Dim sc As IStoredClass = container.Ext().StoredClass(GetType(Car))
                Dim sf As IStoredField = sc.StoredField("_pilot", GetType(Pilot))
                System.Console.WriteLine("Field info:  " + sf.GetName() + "/" + sf.GetStoredType().GetName() + "/IsArray=" + sf.IsArray().ToString())

                System.Console.WriteLine("Retrieve all fields: ")
                Dim sfields() As IStoredField = sc.GetStoredFields()
                Dim i As Integer
                For i = 0 To sfields.Length - 1 Step i + 1
                    System.Console.WriteLine("Stored field:  " + sfields(i).GetName() + "/" + sfields(i).GetStoredType().GetName())
                Next
            Finally
                container.Close()
            End Try
        End Sub
        ' end GetMetaObjectsInfo

    End Class
End Namespace

File: reference/reference/src/com/db4odoc/metainfo/MetaInfoExample.java
/* Copyright (C) 2004 - 2007 db4objects Inc. http://www.db4o.com */

package com.db4odoc.metainfo;

import java.io.File;

import com.db4o.Db4o;
import com.db4o.ObjectContainer;
import com.db4o.ext.StoredClass;
import com.db4o.ext.StoredField;

public class MetaInfoExample {
	private final static String DB4O_FILE_NAME = "reference.db4o";

	public static void main(String[] args) {
		setObjects();
		getMetaObjects();
		getMetaObjectsInfo();
	}

	// end main

	private static void setObjects() {
		new File(DB4O_FILE_NAME).delete();
		ObjectContainer container = Db4o.openFile(DB4O_FILE_NAME);
		try {
			Car car = new Car("BMW", new Pilot("Rubens Barrichello"));
			container.set(car);
			car = new Car("Ferrari", new Pilot("Michael Schumacher"));
			container.set(car);
		} finally {
			container.close();
		}
	}

	// end setObjects

	private static void getMetaObjects() {
		ObjectContainer container = Db4o.openFile(DB4O_FILE_NAME);
		try {
			System.out
					.println("Retrieve meta information for class: ");
			StoredClass sc = container.ext().storedClass(
					Car.class.getName());
			System.out.println("Stored class:  " + sc.toString());

			System.out
					.println("Retrieve meta information for all classes in database: ");
			StoredClass sclasses[] = container.ext().storedClasses();
			for (int i = 0; i < sclasses.length; i++) {
				System.out.println(sclasses[i].getName());
			}
		} finally {
			container.close();
		}
	}

	// end getMetaObjects

	private static void getMetaObjectsInfo() {
		ObjectContainer container = Db4o.openFile(DB4O_FILE_NAME);
		try {
			System.out
					.println("Retrieve meta information for field: ");
			StoredClass sc = container.ext().storedClass(
					Car.class.getName());
			StoredField sf = sc.storedField("pilot", Pilot.class);
			System.out
					.println("Field info:  " + sf.getName() + "/"
							+ sf.getStoredType() + "/isArray="
							+ sf.isArray());

			System.out.println("Retrieve all fields: ");
			StoredField sfields[] = sc.getStoredFields();
			for (int i = 0; i < sfields.length; i++) {
				System.out.println("Stored field:  "
						+ sfields[i].getName() + "/"
						+ sfields[i].getStoredType());
			}
		} finally {
			container.close();
		}
	}
	// end getMetaObjectsInfo
}
